tikv/raftstore/util.go

```diff
@@ -0,0 +1,229 @@
+	LeaseState_Expired
+)
+
+/// Lease records an expired time, for examining the current moment is in lease or not.
+/// It's dedicated to the Raft leader lease mechanism, contains either state of
+///   1. Suspect Timestamp
+///      A suspicious leader lease timestamp, which marks the leader may still hold or lose
+///      its lease until the clock time goes over this timestamp.
+///   2. Valid Timestamp
+///      A valid leader lease timestamp, which marks the leader holds the lease for now.
+///      The lease is valid until the clock time goes over this timestamp.
+///
+/// ```text
+/// Time
+/// |---------------------------------->
+///         ^               ^
+///        Now           Suspect TS
+/// State:  |    Suspect    |   Suspect
+///
+/// |---------------------------------->
+///         ^               ^
+///        Now           Valid TS
+/// State:  |     Valid     |   Expired
+/// ```
+///
+/// Note:
+///   - Valid timestamp would increase when raft log entries are applied in current term.
+///   - Suspect timestamp would be set after the message `MsgTimeoutNow` is sent by current peer.
+///     The message `MsgTimeoutNow` starts a leader transfer procedure. During this procedure,
+///     current peer as an old leader may still hold its lease or lose it.
+///     It's possible there is a new leader elected and current peer as an old leader
+///     doesn't step down due to network partition from the new leader. In that case,
+///     current peer lose its leader lease.
+///     Within this suspect leader lease expire time, read requests could not be performed
+///     locally.
+///   - The valid leader lease should be `lease = max_lease - (commit_ts - send_ts)`
+///     And the expired timestamp for that leader lease is `commit_ts + lease`,
+///     which is `send_ts + max_lease` in short.
+type Lease struct {
+	// If boundSuspect is not nil, then boundValid must be nil, if boundValid is not nil, then
+	// boundSuspect must be nil
+	boundSuspect *time.Time
+	boundValid *time.Time
+	maxLease time.Duration
+
+	maxDrift time.Duration
+	lastUpdate time.Time
+	remote *RemoteLease
+}
+
+func NewLease(maxLease time.Duration) *Lease {
+	return &Lease{
+		maxLease: maxLease,
+		maxDrift: maxLease / 3,
+		lastUpdate: time.Time{},
+	}
+}
+
+/// The valid leader lease should be `lease = max_lease - (commit_ts - send_ts)`
+/// And the expired timestamp for that leader lease is `commit_ts + lease`,
+/// which is `send_ts + max_lease` in short.
+func (l *Lease) nextExpiredTime(sendTs time.Time) time.Time {
+	return sendTs.Add(l.maxLease)
+}
+
+/// Renew the lease to the bound.
+func (l *Lease) Renew(sendTs time.Time) {
+	bound := l.nextExpiredTime(sendTs)
+	if l.boundSuspect != nil || l.boundValid != nil {
+		// Longer than suspect ts or longer than valid ts.
+		if l.boundSuspect != nil && l.boundSuspect.Before(bound) {
+			l.boundSuspect = nil
+			l.boundValid = &bound
+		}
+		if l.boundValid != nil && l.boundValid.Before(bound) {
```

Comment:
If line 88 set the `boundValid`, then `l.boundValid` would not be nil, it's not the logic we want.
I think this match the orginal logic better.
```go
if l.boundSuspect != nil {
    ...
} else if l.boundValid != nil {
    ...
} else {
    ...
}



```